azurerm/internal/services/mssql/mssql_server_transparent_data_encryption_resource.go

```diff
@@ -0,0 +1,213 @@
+package mssql
+
+import (
+	"fmt"
+	"log"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/services/preview/sql/mgmt/v3.0/sql"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/azure"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/clients"
+	keyVaultParser "github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/keyvault/parse"
+	keyVaultValidate "github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/keyvault/validate"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/mssql/parse"
+	azSchema "github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/tf/schema"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/timeouts"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func resourceMsSqlTransparentDataEncryption() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceMsSqlTransparentDataEncryptionCreateUpdate,
+		Read:   resourceMsSqlTransparentDataEncryptionRead,
+		Update: resourceMsSqlTransparentDataEncryptionCreateUpdate,
+		Delete: resourceMsSqlTransparentDataEncryptionDelete,
+
+		Importer: azSchema.ValidateResourceIDPriorToImport(func(id string) error {
+			_, err := parse.EncryptionProtectorID(id)
+
+			return err
+		}),
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(30 * time.Minute),
+			Read:   schema.DefaultTimeout(5 * time.Minute),
+			Update: schema.DefaultTimeout(30 * time.Minute),
+			Delete: schema.DefaultTimeout(30 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"server_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"resource_group_name": azure.SchemaResourceGroupName(),
+			"key_vault_uri": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: keyVaultValidate.NestedItemId,
+			},
+		},
+	}
+}
+
+func resourceMsSqlTransparentDataEncryptionCreateUpdate(d *schema.ResourceData, meta interface{}) error {
+	encryptionProtectorClient := meta.(*clients.Client).MSSQL.EncryptionProtectorClient
+	serverKeysClient := meta.(*clients.Client).MSSQL.ServerKeysClient
+	ctx, cancel := timeouts.ForCreateUpdate(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	log.Printf("[INFO] preparing arguments for MSSQL Encrypted Protector creation.")
+
+	// encryptedProtectorName := d.Get("name").(string)
+	serverName := d.Get("server_name").(string)
+	resGroup := d.Get("resource_group_name").(string)
```

Comment:
Using the ID means we could have:
```suggestion
      serverId, err := parse.ServerId(d.Get("server_id").(string))
      if err != nil {
        return err
      }
```

Reply:
Updated to use the generated server id